ansible_base/oauth2_provider/serializers/application.py

```diff
@@ -0,0 +1,92 @@
+from django.core.exceptions import ObjectDoesNotExist
+from django.utils.translation import gettext_lazy as _
+from oauth2_provider.generators import generate_client_secret
+
+from ansible_base.lib.serializers.common import NamedCommonModelSerializer
+from ansible_base.lib.utils.encryption import ENCRYPTED_STRING
+from ansible_base.oauth2_provider.models import OAuth2Application
+
+
+class OAuth2ApplicationSerializer(NamedCommonModelSerializer):
+    oauth2_client_secret = None
+
+    class Meta:
+        model = OAuth2Application
+        fields = NamedCommonModelSerializer.Meta.fields + [x.name for x in OAuth2Application._meta.concrete_fields]
+        read_only_fields = ('client_id', 'client_secret')
+        read_only_on_update_fields = ('user', 'authorization_grant_type')
+        extra_kwargs = {
+            'user': {'allow_null': True, 'required': False},
+            'organization': {'allow_null': False},
+            'authorization_grant_type': {'allow_null': False, 'label': _('Authorization Grant Type')},
+            'client_secret': {'label': _('Client Secret')},
+            'client_type': {'label': _('Client Type')},
+            'redirect_uris': {'label': _('Redirect URIs')},
+            'skip_authorization': {'label': _('Skip Authorization')},
+        }
+
+    def _get_client_secret(self, obj):
+        request = self.context.get('request', None)
+        try:
+            if obj.client_type == 'public':
+                return None
+            elif request.method == 'POST':
+                if self.oauth2_client_secret is None:
+                    # This should be an impossible case, but...
+                    return ENCRYPTED_STRING
+                # Show the secret, one time, on POST
+                return self.oauth2_client_secret
+            else:
+                return ENCRYPTED_STRING
+        except ObjectDoesNotExist:
+            return ''
+
+    def to_representation(self, instance):
+        # We have to override this because in AbstractCommonModelSerializer, we'll
+        # auto-force all encrypted fields to ENCRYPTED_STRING. Usually that's fine,
+        # but we want to show the client_secret on POST. Ideally we'd just use
+        # get_client_secret() and a SerializerMethodField.
+        ret = super().to_representation(instance)
+        secret = self._get_client_secret(instance)
+        if secret is None:
+            del ret['client_secret']
+        else:
+            ret['client_secret'] = secret
+        return ret
+
+    def _summary_field_tokens(self, obj):
+        token_list = [{'id': x.pk, 'token': ENCRYPTED_STRING, 'scope': x.scope} for x in obj.access_tokens.all()[:10]]
```

Comment:
In review this feels weird. But I guess its what AWX does.

Reply:
Yeah that felt weird to me too. My guess is that it was for performance reasons, but it feels like in practice someone wouldn't have hundreds/thousands of keys anyway. But I kept it because it's what AWX did.